File: WunderNetDev/WunderNode/WunderLayer.cs
/**
 * Author: Corey Wunderlich
 * Date: 3/25/2016
 * Description: The WunderLayer. This handles all of the basic
 * protocol handshaking and packet deserialization. The basic idea is that
 * any node on the network can discover and control/communicate with other nodes.
 * The basic protocol intends for there to be entities that have features. You
 * can discover entities on the network, get their list of features, and control those
 * features in a standard way.
 */
using System;
using System.Collections.Generic;
using System.Collections;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Net.Sockets;
using System.Net;
using WunderNetLayer;

namespace WunderNetNode
{
    public class BasePacketEventArgs : EventArgs
    {
        public BasePacket packet;
    }
    public class StringDataPacketEventArgs : EventArgs
    {
        public StringDataPacket packet;
    }
    public class DescriptionPacketEventArgs : EventArgs 
    {
        public DescriptionPacket packet;
    }
    public class FeatureUpdatePacketEventArgs : EventArgs
    {
        public FeaturePacket packet;
    }
    
    public class WunderLayer
    {
        WunderNet _TheNet;
        public string Identifier;
        public event EventHandler<BasePacketEventArgs> BasePacketReceived;
        public event EventHandler<StringDataPacketEventArgs> StringDataReceived;
        public event EventHandler<DescriptionPacketEventArgs> DescriptionReceived;
        public event EventHandler<FeatureUpdatePacketEventArgs> FeatureUpdateReceived;
        public WunderLayer(string id)
        {
            this.WunderLayerSharedConstructor(id, "", 1000);
        }
        public WunderLayer(string id, string ip, int port)
        {
            this.WunderLayerSharedConstructor(id, ip, port);
        }
        private void WunderLayerSharedConstructor(string id, string ip, int port)
        {
            this.Identifier = id;
            _TheNet = new WunderNet();
            if (ip != "") _TheNet.SetTXPort(ip, port);
            else _TheNet.SetTXPort(port);
            _TheNet.WunderNetEvent += ProcessWunderNet;
            _TheNet.StartListening(port);
            
        }

        public void SendDiscover()
        {
            SendBasePacket(PacketTypes.DISCOVER);
        }
        public void SendDescribe(string receiver)
        {
            SendBasePacket(receiver, PacketTypes.DESCRIBE);
        }
        public void SendStringData(string receiver, string data)
        {
            StringDataPacket wp = new StringDataPacket();
            wp.SenderID = Identifier;
            wp.ReceiverID = receiver;
            wp.PacketType = (Int32)PacketTypes.STRING;
            wp.Data = data;
            _TheNet.SendPacket(wp.GetBytes());
        }
        public void Disconnect()
        {
            SendBasePacket(PacketTypes.OFFLINE);
            _TheNet.StopListening();
        }

        protected void SendOnline()
        {
            SendBasePacket(PacketTypes.ONLINE);
        }
        protected void SendFeatureUpdate(string receiver, string featureName, string s)
        {
            FeaturePacket update = new FeaturePacket(this.Identifier, receiver, featureName, FeatureBaseTypes.STRING);
            update.Data = Encoding.ASCII.GetBytes(s);
            _TheNet.SendPacket(update.GetBytes());
        }
        protected void SendFeatureUpdate(string receiver, string featureName, UInt32 s)
        {
            FeaturePacket update = new FeaturePacket(this.Identifier, receiver, featureName, FeatureBaseTypes.INT);
            update.Data = BitConverter.GetBytes(s);
            _TheNet.SendPacket(update.GetBytes());
        }
        protected void SendFeatureUpdate(string receiver, string featureName, bool s)
        {
            FeaturePacket update = new FeaturePacket(this.Identifier, receiver, featureName, FeatureBaseTypes.BOOL);
            update.Data = BitConverter.GetBytes(s);
            _TheNet.SendPacket(update.GetBytes());
        }
        protected void SendFeatureSubscribe(string receiver, string featureName)
        {
            FeaturePacket subs = new FeaturePacket(this.Identifier, receiver, featureName, PacketTypes.SUBSCRIBE);
            _TheNet.SendPacket(subs.GetBytes());
        }


        protected void SendDescription(string receiver, StandardFeature[] sf)
        {
            DescriptionPacket dp = new DescriptionPacket();
            dp.SenderID = Identifier;
            dp.ReceiverID = receiver;
            dp.PacketType = (Int32)PacketTypes.DESCRIPTION;
            foreach (StandardFeature f in sf)
            {
                dp.AddFeature(f);
            }
            _TheNet.SendPacket(dp.GetBytes());
        }
        protected void SendDescription(string receiver, ArrayList sf)
        {
            DescriptionPacket dp = new DescriptionPacket();
            dp.SenderID = Identifier;
            dp.ReceiverID = receiver;
            dp.PacketType = (Int32)PacketTypes.DESCRIPTION;
            foreach (StandardFeature f in sf)
            {
                dp.AddFeature(f);
            }
            _TheNet.SendPacket(dp.GetBytes());
        }
        protected void SendDescription(string receiver, ICollection sf)
        {
            DescriptionPacket dp = new DescriptionPacket();
            dp.SenderID = Identifier;
            dp.ReceiverID = receiver;
            dp.PacketType = (Int32)PacketTypes.DESCRIPTION;
            foreach (StandardFeature f in sf)
            {
                dp.AddFeature(f);
            }
            _TheNet.SendPacket(dp.GetBytes());
        }
        protected void SendIdentify(string receiver)
        {
            SendBasePacket(receiver, PacketTypes.IDENTIFY);
        }
        protected void SendBasePacket(PacketTypes type)
        {
            SendBasePacket("", type);
        }
        protected void SendBasePacket(string receiver, PacketTypes type)
        {
            BasePacket wp = new BasePacket();
            wp.SenderID = Identifier;
            wp.ReceiverID = receiver;
            wp.PacketType = (Int32)type;
            _TheNet.SendPacket(wp.GetBytes());
        }

        private void ProcessWunderNet(object sender, WunderNetPacketEventArgs e)
        {
            if(BasePacket.CheckHeader(e.RawData))
            {
                BasePacket bp = new BasePacket(e.RawData);

                if (bp.SenderID == this.Identifier) return;

                switch((PacketTypes)bp.PacketType)
                {
                    case PacketTypes.DATABLOCK: ProcessDataBlock(bp, e.RawData); break;
                    case PacketTypes.STRING: ProcessDataBlock(bp, e.RawData); break;
                    case PacketTypes.DISCOVER: ProcessDiscover(bp); break;
                    case PacketTypes.IDENTIFY: ProcessIdentify(bp); break;
                    case PacketTypes.ONLINE: ProcessBasePacketCallbacks(bp); break;
                    case PacketTypes.OFFLINE: ProcessBasePacketCallbacks(bp); break;
                    case PacketTypes.DESCRIBE: ProcessDescribe(bp); break;
                    case PacketTypes.DESCRIPTION: ProcessDecription(bp, e.RawData); break;
                    case PacketTypes.UPDATE: ProcessFeatureUpdate(bp, e.RawData); break;
                    case PacketTypes.COMMAND: ProcessBasePacketCallbacks(bp); break;
                    case PacketTypes.SUBSCRIBE: ProcessSubscribe(bp, e.RawData); break;
                }

            }
        }

        protected void ProcessBasePacketCallbacks(BasePacket p)
        {
            Delegate[] registeredEvents = BasePacketReceived.GetInvocationList();
            BasePacketEventArgs e = new BasePacketEventArgs();
            e.packet = p;
            foreach (Delegate el in registeredEvents)
            {
                ((EventHandler<BasePacketEventArgs>)el).Invoke(this, e);
            }
        }
        protected void ProcessDescriptionPacketCallbacks(DescriptionPacket p)
        {
            Delegate[] registeredEvents = DescriptionReceived.GetInvocationList();
            DescriptionPacketEventArgs e = new DescriptionPacketEventArgs();
            e.packet = p;
            foreach (Delegate el in registeredEvents)
            {
                ((EventHandler<DescriptionPacketEventArgs>)el).Invoke(this, e);
            }
        }
        protected void ProcessStringDataPacketCallbacks(StringDataPacket p)
        {
            Delegate[] registeredEvents = StringDataReceived.GetInvocationList();
            StringDataPacketEventArgs e = new StringDataPacketEventArgs();
            e.packet = p;
            foreach (Delegate el in registeredEvents)
            {
                ((EventHandler<StringDataPacketEventArgs>)el).Invoke(this, e);
            }
        }
        protected void ProcessFeatureUpdateCallbacks(FeaturePacket p)
        {
            Delegate[] registeredEvents = FeatureUpdateReceived.GetInvocationList();
            FeatureUpdatePacketEventArgs e = new FeatureUpdatePacketEventArgs();
            e.packet = p;
            foreach (Delegate el in registeredEvents)
            {
                ((EventHandler<FeatureUpdatePacketEventArgs>)el).Invoke(this, e);
            }
        }

        protected void ProcessDataBlock(BasePacket bp, byte[] rawBytes)
        {
            if (bp.ReceiverID == this.Identifier)
            {
                StringDataPacket sdp = new StringDataPacket(rawBytes);
                ProcessStringDataPacketCallbacks(sdp);
            }
        }
        protected void ProcessDiscover(BasePacket bp)
        {
            SendIdentify(bp.SenderID);
        }
        protected virtual void ProcessDescribe(BasePacket bp) { }
        protected virtual void ProcessSubscribe(BasePacket bp, byte[] rawBytes) { }
        protected void ProcessIdentify(BasePacket bp)
        {
            if (bp.ReceiverID == this.Identifier) ProcessBasePacketCallbacks(bp);
        }
        protected void ProcessDecription(BasePacket bp, byte[] rawBytes)
        {
            if (bp.ReceiverID == this.Identifier)
            {
                DescriptionPacket dp = new DescriptionPacket(rawBytes);
                ProcessDescriptionPacketCallbacks(dp);
            }
        }
        protected virtual void ProcessFeatureUpdate(BasePacket bp, byte[] rawBytes)
        {
            if (bp.ReceiverID == this.Identifier)
            {
                FeaturePacket dp = new FeaturePacket(rawBytes);
                ProcessFeatureUpdateCallbacks(dp);
            }
        }

    }
}

File: WunderNodePython/WunderLayer.py
#!/usr/bin/env python
# This is WunderLayer Python Edition

import WunderNet
import WunderPackets

class WunderLayer:
	def __init__(self, id, baddr):
		self.Identifier = id

		self.TheNet = WunderNet.WunderNet(baddr)
		self.TheNet.RegisterCallback(self.ProcessWunderNet)
		self.TheNet.StartListening(1000)
		
		self._BasePacketEventListeners = []
		self._StringDataPacketEventListeners = []
		self._DescriptionPacketEventListeners = []
		self._UpdatePacketEventListeners = []
		self._CommandPacketEventListeners = []
		
		self._packetFunctions = { 
			WunderPackets.PacketTypes['ONLINE'] : self.ProcessBasePacket,
			WunderPackets.PacketTypes['OFFLINE'] : self.ProcessBasePacket,
			WunderPackets.PacketTypes['DATABLOCK'] : self.ProcessDataBlock,
			WunderPackets.PacketTypes['DISCOVER'] : self.ProcessDiscover,
			WunderPackets.PacketTypes['IDENTIFY'] : self.ProcessBasePacket,
			WunderPackets.PacketTypes['DESCRIBE'] : self.ProcessDescribe,
			WunderPackets.PacketTypes['DESCRIPTION'] : self.ProcessDescription,
			WunderPackets.PacketTypes['UPDATE'] : self.ProcessFeatureUpdate,
			WunderPackets.PacketTypes['COMMAND'] : self.ProcessFeatureCommand,
			WunderPackets.PacketTypes['SUBSCRIBE'] : self.ProcessSubscribe
		}
		
	def RegisterForBasePackets(self, function):
		self._BasePacketEventListeners.append(function)
	
	def RegisterForStringDataPackets(self, function):
		self._StringDataPacketEventListeners.append(function)
		
	def RegisterForDescriptionPackets(self, function):
		self._DescriptionPacketEventListeners.append(function)
	
	def RegisterForFeaturePackets(self, function):
		self._UpdatePacketEventListeners.append(function)
		
	def RegisterForFeatureCommands(self, function):
		self._CommandPacketEventListeners.append(function)
		
	def AddFeature(self, feature):
		self.FeatureList.append(feature)
	
	def SendOnline(self):
		p = WunderPackets.BasePacket()
		p.InitPacket(self.Identifier, '', WunderPackets.PacketTypes['ONLINE'])
		self.TheNet.SendPacket(p.GetBytes())
	
	def SendOffline(self):
		p = WunderPackets.BasePacket()
		p.InitPacket(self.Identifier, '', WunderPackets.PacketTypes['OFFLINE'])
		self.TheNet.SendPacket(p.GetBytes())
	
	def SendIdentify(self, receiver):
		p = WunderPackets.BasePacket()
		p.InitPacket(self.Identifier, receiver, WunderPackets.PacketTypes['IDENTIFY'])
		self.TheNet.SendPacket(p.GetBytes())
		
	def SendDesciption(self, receiver, features):
		p = WunderPackets.DescriptionPacket()
		p.InitPacket(self.Identifier, receiver, features)
		self.TheNet.SendPacket(p.GetBytes())
	
	def SendFeatureUpdate(self, receiver, fname, ftype, fdata):
		p = WunderPackets.FeaturePacket()
		p.InitPacket(self.Identifier, receiver, WunderPackets.PacketTypes['UPDATE'],fname,ftype,fdata)
		self.TheNet.SendPacket(p.GetBytes())
		
	def SendFeatureCommand(self, receiver, fname, ftype, fdata):
		p = WunderPackets.FeaturePacket()
		p.InitPacket(self.Identifier, receiver, WunderPackets.PacketTypes['COMMAND'],fname,ftype,fdata)
		self.TheNet.SendPacket(p.GetBytes())	
	
	def SendFeatureSubscribe(self, receiver, fname):
		p = WunderPackets.FeaturePacket()
		p.InitPacket(self.Identifier, receiver, WunderPackets.PacketTypes['SUBSCRIBE'],fname,0,'')
		self.TheNet.SendPacket(p.GetBytes())
	
	def Disconnect(self):
		self.SendOffline()
		self.TheNet.StopListening()
	
	def ProcessWunderNet(self,data,*args):
		if WunderPackets.CheckHeader(data):
			# print 'WunderData Found'
			p = WunderPackets.BasePacket()
			p.InitFromPacket(data)
			if p.SenderID == self.Identifier:
				return
			# print p.SenderID + ":" + str(p.PacketType)	
			self._packetFunctions[p.PacketType](p, data);
	
	def ProcessBasePacket(self, wpacket, rawdata):
		for callback in self._BasePacketEventListeners:
			callback(wpacket)
	
	def ProcessDiscover(self, wpacket, rawdata):
		self.SendIdentify(wpacket.SenderID)
	
	def ProcessDataBlock(self, wpacket, rawdata):
		if wpacket.ReceiverID == self.Identifier:
			dBlock = WunderPackets.StringDataPacket()
			dBlock.InitFromPacket(rawdata)
			for callback in self._StringDataPacketEventListeners:
				callback(dBlock)
	def ProcessDescribe(self, wpacket, rawdata):
		return False
	
	def ProcessSubscribe(self, wpacket, rawdata):
		return False
		
	def ProcessDescription(self, wpacket, rawdata):
		if wpacket.ReceiverID == self.Identifier:
			for callback in self._DescriptionPacketEventListeners:
				callback(wpacket)
				
	def ProcessFeatureUpdate(self, wpacket, rawdata):
		if wpacket.ReceiverID == self.Identifier:
			uBlock = WunderPackets.FeaturePacket()
			uBlock.InitFromPacket(rawdata)
			for callback in self._UpdatePacketEventListeners:
				callback(uBlock)
				
	def ProcessFeatureCommand(self, wpacket, rawdata):
		if wpacket.ReceiverID == self.Identifier:
			uBlock = WunderPackets.FeaturePacket()
			uBlock.InitFromPacket(rawdata)
			for callback in self._CommandPacketEventListeners:
				callback(uBlock)

